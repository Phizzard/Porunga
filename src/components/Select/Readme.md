Default Select

```jsx
<Select
  options={[
    { value: 'option-1', label: 'Option 1' },
    { value: 'option-2', label: 'Option 2' },
  ]}
/>
```

Default Select

```jsx
<Select
  defaultLabel="Kept you waiting huh?"
  options={[
    { value: 'option-1', label: 'Option 1' },
    { value: 'option-2', label: 'Option 2' },
  ]}
/>
```

Select Palette

```jsx
<Select
  primaryColor="brand"
  options={[
    { value: 'option-1', label: 'Option 1' },
    { value: 'option-2', label: 'Option 2' },
  ]}
/>
<Select
  primaryColor="accent-1"
  options={[
    { value: 'option-1', label: 'Option 1' },
    { value: 'option-2', label: 'Option 2' },
  ]}
/>
<Select
  primaryColor="success"
  options={[
    { value: 'option-1', label: 'Option 1' },
    { value: 'option-2', label: 'Option 2' },
  ]}
/>
<Select
  primaryColor="warning"
  options={[
    { value: 'option-1', label: 'Option 1' },
    { value: 'option-2', label: 'Option 2' },
  ]}
/>
<Select
  primaryColor="error"
  options={[
    { value: 'option-1', label: 'Option 1' },
    { value: 'option-2', label: 'Option 2' },
  ]}
/>
```

Select Palette Rounded

```jsx
<Select
  rounded
  primaryColor="brand"
  options={[
    { value: 'option-1', label: 'Option 1' },
    { value: 'option-2', label: 'Option 2' },
  ]}
/>
<Select
  rounded
  primaryColor="accent-1"
  options={[
    { value: 'option-1', label: 'Option 1' },
    { value: 'option-2', label: 'Option 2' },
  ]}
/>
<Select
  rounded
  primaryColor="success"
  options={[
    { value: 'option-1', label: 'Option 1' },
    { value: 'option-2', label: 'Option 2' },
  ]}
/>
<Select
  rounded
  primaryColor="warning"
  options={[
    { value: 'option-1', label: 'Option 1' },
    { value: 'option-2', label: 'Option 2' },
  ]}
/>
<Select
  rounded
  primaryColor="error"
  options={[
    { value: 'option-1', label: 'Option 1' },
    { value: 'option-2', label: 'Option 2' },
  ]}
/>
```
