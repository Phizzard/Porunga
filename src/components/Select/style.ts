import styled from '@emotion/styled'
import { SerializedStyles } from '@emotion/core'

interface SelectProps {
  primaryColor?: string
  secondaryColor?: string
  focusOutline?: string
  shadow?: string
  rounded: SerializedStyles
}

export const Container = styled.select<SelectProps>`
  padding: 0.2rem 0.4rem;
  font-size: 1rem;
  border-width: 0.1rem;
  border-color: ${p => p.primaryColor};
  color: ${p => p.primaryColor};
  background-color: ${p => p.secondaryColor};
  box-shadow: ${p => p.shadow};
  ${p => p.rounded};
  &:focus {
    outline: ${p => p.focusOutline} auto 1px;
    background-color: ${p => p.primaryColor}30;
  }
`

export const Option = styled.option``
