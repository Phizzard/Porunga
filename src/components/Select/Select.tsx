import React from 'react'
import { useTheme } from 'emotion-theming'

import { ThemeConfig, Rounded, Sizes, Palette } from '../../types'
import { roundedStyles } from '../../utils/roundedStyles'
import { get } from '../../utils/get'
import { Container, Option } from './style'

export interface SelectProps {
  /** option elements inside Select {label: string, value: string, selected?: boolean, disabled?: boolean} */
  options: Array<Option>
  /** The default label selected */
  defaultLabel?: string
  /** Whether to round the edges based off of the theme config */
  rounded?: Rounded
  /** The amount of box shadow applied to box */
  shadow?: Sizes
  /** The primary color to use from the theme palette */
  primaryColor?: Palette
  /** The secondary color to use from the theme palette */
  secondaryColor?: Palette
  /** The interface for capturing values of selected option during the select onChange event */
  onChange?: (change: {
    option: { value: string | number; label: string | null }
  }) => any
}

export interface Option {
  label: string
  value: string
  selected?: boolean
  disabled?: boolean
}

const Select = ({
  options,
  rounded,
  onChange,
  shadow = 'unset',
  defaultLabel = 'Please Select',
  primaryColor = 'dark-1',
  secondaryColor = 'light-1',
}: SelectProps) => {
  const theme: ThemeConfig = useTheme()
  const roundedProps = roundedStyles(theme, rounded, 'select')
  const selectedProps = {
    primaryColor: get(theme, 'root', 'colors', primaryColor),
    secondaryColor: get(theme, 'root', 'colors', secondaryColor),
    focusOutline: get(theme, 'root', 'colors', 'accent-1'),
    shadow: get(theme, 'root', 'shadows', shadow),
    rounded: roundedProps,
  }

  return (
    <Container {...selectedProps} onChange={handleOnChange}>
      <Option value="" disabled selected>
        {defaultLabel}
      </Option>
      {options.map(({ value, label }) => (
        <Option key={value} value={value}>
          {label}
        </Option>
      ))}
    </Container>
  )

  function handleOnChange(event: React.ChangeEvent<HTMLSelectElement>) {
    if (event && onChange) {
      const index = event.currentTarget.selectedIndex
      const change = {
        option: {
          value: event.currentTarget.value,
          label: event.currentTarget[index].textContent,
        },
      }

      onChange(change)
    }
  }
}

export default Select
