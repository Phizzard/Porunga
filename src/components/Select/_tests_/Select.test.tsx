import React from 'react'
import Select from '../Select'
import renderer from 'react-test-renderer'

it('renders correctly', () => {
  const component = renderer.create(
    <Select options={[{ label: 'a label', value: 'some value' }]} />
  )

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
