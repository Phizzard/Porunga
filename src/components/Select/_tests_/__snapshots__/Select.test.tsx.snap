// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<select
  className="css-4xqwg4"
  onChange={[Function]}
>
  <option
    className="css-0"
    disabled={true}
    selected={true}
    value=""
  >
    Please Select
  </option>
  <option
    className="css-0"
    value="some value"
  >
    a label
  </option>
</select>
`;
